packages/x-data-grid/src/hooks/features/rowSelection/useGridRowSelection.ts

```diff
@@ -620,14 +638,33 @@ export const useGridRowSelection = (
     GridEventListener<'headerSelectionCheckboxChange'>
   >(
     (params) => {
-      const rowsToBeSelected =
-        props.pagination && props.checkboxSelectionVisibleOnly && props.paginationMode === 'client'
-          ? gridPaginatedVisibleSortedGridRowIdsSelector(apiRef)
-          : gridExpandedSortedRowIdsSelector(apiRef);
-
-      apiRef.current.selectRows(rowsToBeSelected, params.value);
+      if (params.value) {
+        if (!props.isRowSelectable && !props.checkboxSelectionVisibleOnly && applyAutoSelection) {
+          apiRef.current.selectRows({ type: 'exclude', ids: new Set() }, params.value, true);
+        } else {
```

Comment:
The more corner cases I fix, the fewer opportunities for the `exclude` selection model 😅

Reply:
AFAIU it would only be for "select all", but for large datasets it would improve that interaction substantially.

Reply:
Correct, but there is a number of conditions for the exclude model to be used:

```ts
if (
  !props.isRowSelectable &&
  !props.checkboxSelectionVisibleOnly &&
  applyAutoSelection &&
  !hasFilters
) {
  apiRef.current.selectRows({ type: 'exclude', ids: new Set() }, params.value, true);
} else {
  const rowsToBeSelected = getRowsToBeSelected();
  apiRef.current.selectRows(
    { type: 'include', ids: new Set(rowsToBeSelected) },
    params.value,
  );
}
```

Reply:
Ok, so `{ type: 'exclude', ids: new Set() }` could be interpreted differently:
1. All rows are selected (current approach)
2. All selectable and visible rows are selected (this is what you meant, right?)

Do you see benefits in (2) compared to (1)?

Reply:
Yes, I guess 2 would remove all those exceptions. From what I can see, we don't currently support keeping non-visible rows selected, from what I can test [on this example](https://stackblitz.com/edit/react-b717a7?file=Demo.tsx).

https://github.com/user-attachments/assets/7ecc482a-9aca-45e6-a429-b276937a07cb

Reply:
> From what I can see, we don't currently support keeping non-visible rows selected

We do support that with the `keepNonExistentRowsSelected` flag: https://mui.com/x/react-data-grid/row-selection/#row-selection-with-filtering

Reply:
On another thought, there's a difference between (1) and (2) when interpreting this model by the developers to perform operations on selected rows.
In (1), an empty exclude model means literally all rows.
In (2), an empty exclude model could mean different things depending on the data grid state. So developers would need to convert empty exclude into include (?) model to be able to perform actions on the dataset.

With the above, I think we should proceed with (1).
What do you think?

Reply:
Yes, I think (1) is the way to go. Both shapes of the model should be equivalent, which isn't the case for (2).

Reply:
> The more corner cases I fix, the fewer opportunities for the exclude selection model 

It's not that it's not usable for the conditions that you've guarded for, it's that some of them might require building a non-empty exclude model. You'd need an equivalent `getRowsToBeSelectedAsExclude()` function. And there are some cases in which it could make sense to do that work. For example, if there are filters applied that show 10 out of 1,000,000 total rows, using an exclude model would be much faster than building an include one.